core/src/main/java/io/questdb/cairo/CairoEngine.java

```diff
@@ -151,74 +158,160 @@ public void close() {
+            tableNameRegistry.unlockTableName(tableToken);
         }
+        return tableToken;
     }
 
     // caller has to acquire the lock before this method is called and release the lock after the call
     public void createTableUnsafe(
             CairoSecurityContext securityContext,
             MemoryMARW mem,
             Path path,
-            TableStructure struct
+            TableStructure struct,
+            TableToken tableToken,
+            int tableId
     ) {
         securityContext.checkWritePermission();
-        int tableId = (int) tableIdGenerator.getNextId();
-        if (struct.isWalEnabled()) {
-            tableSequencerAPI.registerTable(tableId, struct);
-        }
 
         // only create the table after it has been registered
+        final FilesFacade ff = configuration.getFilesFacade();
+        final CharSequence root = configuration.getRoot();
+        final int mkDirMode = configuration.getMkDirMode();
         TableUtils.createTable(
-                configuration,
+                ff,
+                root,
+                mkDirMode,
                 mem,
                 path,
+                tableToken.getDirName(),
                 struct,
+                ColumnType.VERSION,
                 tableId
         );
+        if (struct.isWalEnabled()) {
+            tableSequencerAPI.registerTable(tableId, struct, tableToken);
+        }
+    }
+
+    public void drop(
+            CairoSecurityContext securityContext,
+            Path path,
+            TableToken tableToken
+    ) {
+        securityContext.checkWritePermission();
+        verifyTableToken(tableToken);
+        if (tableToken.isWal()) {
+            if (tableNameRegistry.dropTable(tableToken)) {
+                tableSequencerAPI.dropTable(tableToken, false);
+            } else {
+                LOG.info().$("table is already dropped [table=").$(tableToken)
+                        .$(", dirName=").$(tableToken.getDirName()).I$();
+            }
+        } else {
+            CharSequence lockedReason = lock(securityContext, tableToken, "removeTable");
+            if (null == lockedReason) {
+                try {
+                    path.of(configuration.getRoot()).concat(tableToken).$();
+                    int errno;
+                    if ((errno = configuration.getFilesFacade().rmdir(path)) != 0) {
+                        LOG.error().$("drop failed [tableName='").$(tableToken).$("', error=").$(errno).$(']').$();
+                        throw CairoException.critical(errno).put("could not remove table [name=").put(tableToken)
+                                .put(", dirName=").put(tableToken.getDirName()).put(']');
+                    }
+                } finally {
+                    readerPool.unlock(tableToken);
+                    writerPool.unlock(tableToken, null, false);
+                    metadataPool.unlock(tableToken);
+                }
+
+                tableNameRegistry.dropTable(tableToken);
```

Comment:
This call includes an unconditional removal from CHM:
```java
return reverseTableNameTokenMap.remove(token.getDirName()) != null;
```

Since we already unlocked all pools at this point, nothing prevents a concurrent `createTable` invocation to execute `registerName` and do an unconditional put on the same CHM key:
```java
reverseTableNameTokenMap.put(tableToken.getDirName(), ReverseTableMapItem.of(tableToken));
```

As a result, the original `drop` call will remove the entry from `reverseTableNameTokenMap` for a just created table. To fix that, we need to move this call into the above `try` block.

The same consideration applies to `rename`.

Reply:
> This call includes an unconditional removal from CHM:
> 
> ```java
> return reverseTableNameTokenMap.remove(token.getDirName()) != null;
> ```

It is not unconditional, it happens when `nameTableTokenMap` removal was successful:

```
        if (nameTableTokenMap.remove(token.getTableName(), token)) {
            ...
            return reverseTableNameTokenMap.remove(token.getDirName()) != null;
        }
```

Create table reserves table name first in `nameTableTokenMap`.
Generally `reverseTableNameTokenMap` is not racy, races are checked in `nameTableTokenMap`, reverse map only follows.

Reply:
Changed to conditional remove in [1abaa43](https://github.com/questdb/questdb/pull/2752/commits/1abaa431fd00b2ab1adc0d87563795ce33f228a1)